src/t_hash.c

```diff
@@ -29,52 +29,183 @@
+         *     | ptr   | hdr8 "foo" \0 |
+         *     +-------+---------------+
+         */
+        char field_sds_type = sdsReqType(field_len);
+        if (field_sds_type == SDS_TYPE_5) field_sds_type = SDS_TYPE_8;
+        field_size = sdsReqSize(field_len, field_sds_type);
+        size_t alloc_size = sizeof(sds *) + field_size;
+        char *buf = zmalloc(alloc_size);
+        *(sds *)buf = value;
+        embedded_field_sds = sdswrite(buf + sizeof(sds *), field_size, field_sds_type, field, field_len);
+        /* Store the entry encoding type in sds aux bits. */
+        sdsSetAuxBit(embedded_field_sds, FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR, 1);
+        serverAssert(entryHasValuePtr(embedded_field_sds));
+    }
+    return (void *)embedded_field_sds;
 }
 
+/* The entry pointer is the field sds, but that's an implementation detail. */
 sds hashTypeEntryGetField(const hashTypeEntry *entry) {
-    const char *field = entry->field_data + entry->field_offset;
-    return (sds)field;
+    return (sds)entry;
 }
 
 sds hashTypeEntryGetValue(const hashTypeEntry *entry) {
-    return entry->value;
+    if (entryHasValuePtr(entry)) {
+        return *hashTypeEntryGetValueRef(entry);
+    } else {
+        /* Skip field content, field null terminator and value sds8 hdr. */
+        size_t offset = sdslen(entry) + 1 + sdsHdrSize(SDS_TYPE_8);
+        char *buf = (char *)entry + offset;
+        return buf;
+    }
 }
 
-/* frees previous value, takes ownership of new value */
-static void hashTypeEntryReplaceValue(hashTypeEntry *entry, sds value) {
-    sdsfree(entry->value);
-    entry->value = value;
+/* Returns the address of the entry allocation. */
+static void *hashTypeEntryAllocPtr(hashTypeEntry *entry) {
+    char *buf = sdsAllocPtr(entry);
+    if (entryHasValuePtr(entry)) {
+        buf -= sizeof(sds *);
+    }
+    return buf;
 }
 
-/* Returns allocation size of hashTypeEntry and data owned by hashTypeEntry,
- * even if not embedded in the same allocation. */
-size_t hashTypeEntryAllocSize(hashTypeEntry *entry) {
-    size_t size = zmalloc_usable_size(entry);
-    size += sdsAllocSize(entry->value);
-    return size;
+/* Frees previous value, takes ownership of new value, returns entry (may be
+ * reallocated). */
+static hashTypeEntry *hashTypeEntryReplaceValue(hashTypeEntry *entry, sds value) {
+    sds field = (sds)entry;
+    size_t field_size = sdsHdrSize(sdsType(field)) + sdsalloc(field) + 1;
+    size_t value_len = sdslen(value);
+    size_t value_size = sdsReqSize(value_len, SDS_TYPE_8);
+    if (!entryHasValuePtr(entry)) {
+        /* Reuse the allocation if the new value fits and leaves no more than
+         * 25% unused space after replacing the value. */
+        char *alloc_ptr = sdsAllocPtr(entry);
+        size_t required_size = field_size + value_size;
+        size_t alloc_size;
+        if (required_size <= EMBED_VALUE_MAX_ALLOC_SIZE &&
+            required_size <= (alloc_size = zmalloc_usable_size(alloc_ptr)) &&
```

Comment:
Just a small note. I prefer this implementation because of the simplicity.
I would say that we could only compare the value size diff (the current VS the new value) since the field is immutable and the size is only impacted by the values diff. Only stating this since IIRC zmalloc_usable_size has some notable performance impact

Reply:
I managed to get rid of zmalloc_usable_size. We encode the unused space in the embedded value SDS header's alloc field. It can be used.